### DISCUSSION TITLE
Host the Digital Pali Dictionary Online

### DISCUSSION THREAD

**Discussion by Devamitta** [digitalpalidictionary/dpd-db#7] [Ideas]
I wanted to share an idea that could greatly enhance the ease of use and accessibility of this valuable resource.

### The Idea:
Host the DPD online, making it available to anyone with internet access without the need for installation.

### Rationale:

**Universal Access:** Users can access the DPD from any device and location, provided they have an internet connection. This is especially beneficial for those who may not have the technical expertise or resources to install the software.

**Continuous Updates:** An online platform could allow for easier and more frequent updates, ensuring users always have access to the latest content.

**Reach a Broader Audience:** While DPD already has a committed user base, hosting it online can make it available to an even wider audience, including those who might stumble upon it without prior knowledge of its existence.

### In Conclusion:
Having the DPD hosted online could greatly benefit our existing user base and potentially introduce this rich resource to many more individuals around the globe. I believe in the potential of this idea and would love to hear feedback or any offers of support.

Thank you for considering this proposal. Looking forward to the continued growth and success of the Digital Pali Dictionary.

**Comment by gambhiro** [comment]
One way I could see how to do this Ven. @Devamitta , is a small frontend interface making Ajax fetch queries to the SQLite db, or better, the same imported to MySQL.

Connecting to SQLite can be slow, because reading/writing to disk in hosted environments is even worse than on your desktop.

Dropping the non-essential tables can bring down the size, and then an SQL dump can be directly imported into MySQL. Since there is no user data, you could update the db by dropping the tables and re-importing everything.

Shared hosting usually only offers PHP, which is a horror to code in.

I can recommend PythonAnywhere though: https://www.pythonanywhere.com/

Even the free tier account would be sufficient to run this. They have a detailed tutorial for creating a small Flask server with a db backend: [part 1](https://blog.pythonanywhere.com/121/) and [part 2](https://blog.pythonanywhere.com/158/). See also [this](https://help.pythonanywhere.com/pages/UsingSQLAlchemywithMySQL) and [this](https://help.pythonanywhere.com/pages/UsingMySQL).

For frontend code I would recommend VueJS https://vuejs.org/

It's basically a more simple ReactJS from ppl who found that over-complicated. I used it for this: https://ticivara.github.io/en/sabong-pattern

**Reply by falko-strenzke** [reply]
My thought was to parse the MDict contents using https://github.com/binhetech/mdict-parser/ (I already check that it works on DPD), put everything into database and write a simple server that servers the entries. 

There is only very little additional work necessary as it seems to me currently:
- resolve the "link" entries for inflected to point to the right headword
- make every dictionary word within the entries a link
- find some suitable way of data compression. Since we have the case of repeated text strings of considerable length, a very simple approach can be taken that is based on extracting a template (which can be just one arbitrary dictionary entry page that contains all the types of fields) the contents of which are then references by the entries. This will probably decrease the amound of data to a  reasonable size already (simply extracting all the MDict entries into one directory gave me 7.2 GB so far).
- store all the entries in two tables:
  - headword table with the actual entries
  - inflection table with links to the headwords (and ideally already the indications of the inflected form)

Regarding the speed of SQLite in a hosted environment I am not experienced, but I would tend to give it a try first. Regarding the programming language and hosting evironment, I would definitely not use Python but code it natively based on suitable industry qualilty web framework. I am prepared to take over the hosting cost for a [V-Server, which is available from 6€ / month](https://www.strato.de/server/linux-vserver/). 

I completely understand the prevelance of Python Web Frameworks and know that they are also widespread in industry applications. It is just my personal experience that working with Python is simple in the beginning, but as the complexity of the code base grows, makes it very hard to keep an overview and consistency. I think what I described is such a small task it can easily be done by one person. Afterwards various changes and additions are possible, like basing everything on a special export format from DPD and a JS frontend. Many improvements regarding options for search queries and display of search results will then suggest themselves.

**Reply by bdhrs** [reply]
Yes, I like the idea of get something working and then improve on it.

Please make sure you understand this flow-chart, it contains everything necessary to display the correct headwords for any inflected form. https://github.com/digitalpalidictionary/dpd-db/tree/main/tbw

> I would definitely not use Python but code it natively based on suitable industry qualilty web framework.

What language are you thinking of working in? There are so many readily available functions and resources for this project in Python, it seems odd to reinvent the wheel.

**Reply by falko-strenzke** [reply]
> Please make sure you understand this flow-chart, it contains everything necessary to display the correct headwords for any inflected form. https://github.com/digitalpalidictionary/dpd-db/tree/main/tbw

I'm not sure that is necessarily already now applicable. To start with, the behaviour of GoldenDict could be emulated, which means just look up the word in any table and display all the results. But regarding that flowchart: What are the .json files you mention there? Are they already created? Can I see (an excerpt of) them?

> What language are you thinking of working in? There are so many readily available functions and resources for this project in Python, it seems odd to reinvent the wheel.

I would program it in C/C++. Create a C++ adaptor for the HTTP server framework https://mongoose.ws/ and otherwise use only C++. I am not sure if there is so much overlap in the functionality. The server application I intend to create would be for the display of the DPD, and not try to take over any tasks in the creation/processing of the data. In any case, it will serve the performance if there is a huge database with all the inflected forms for simple lookup instead of performing any processing in the server. So I am not sure if there is a large potential for overlap in functionality with the existing DPD code. One exception would be the HTML rendering, as soon as we shift from storing plain HTML in the database to a structured approach. But to write the HTML rendering anew doesn't seem to be an overly complicated task.

**Comment by falko-strenzke** [comment]
So actually I gave it a try with Django: https://github.com/falko-strenzke/dpd-server-django

It works almost as simple as expected. The word lookup seems to work well. The search page would need a lot of improvement. One limit here is currently the Django framework which apparently requires tweaks in your code in order to allow for pagination together with filtering. This is in line with my previous experience when I was working with Django that as soon as you go a little bit beyond the absolute basics you have to start working with tweaks and things tend to get dirty. This is the reason why I prefer to write all this search functionality from scratch in a clean manner in a strongly typed language.

But on the other hand the modern way to implement all this would be to just implement a Json API and an application based on a JavaScript framework. The problem here is that is completely outside my skills.

**Reply by bdhrs** [reply]
Great, send a link when you're happy. The next release is the full moon on 27 November, do you think you'll have a public version before then?

**Reply by falko-strenzke** [reply]
Probably not until that date, I am quite busy the next weeks. It is rather likely that I will only manage to complete it in January.

**Reply by falko-strenzke** [reply]
I have the server up now: http://niyamata.de/dict/dpd/search/ 

No HTTPS so far, but will come soon. Please let me know if you think anything should be corrected.

**Reply by bdhrs** [reply]
Ah I'm sorry I didn't catch this sooner, I've just sent out the mailer.

I have one suggestion to replace the word MDict universally with the website name or something relevant. That will let me know on all feedback links where the feedback is coming from.

**Reply by bdhrs** [reply]
There's one other obvious imrpovement within the current structure, and that is to move the <style> for every word into CSS, and the <script> into the page itself, that will reduce your file sizes enormously.

**Reply by falko-strenzke** [reply]
> I have one suggestion to replace the word MDict universally with the website name or something relevant. That will let me know on all feedback links where the feedback is coming from.

Yes, that will be easy to incorporate into the MDict parsing script. I will implement that when I update to the new DPD release. Hopefully next weekend, but I cannot promise it...

**Reply by falko-strenzke** [reply]
I now replaced "MDict" with "niyamata.de" when updating to the Nov. release.

**Comment by Devamitta** [comment]
It is great!

Is it possible to create hyperlinks for searching all Pāli words in the results? Similar to the Golden Dictionary, where you can click on any word and see its corresponding result. 

For instance, if I want to search for variant Pāli words:
![image](https://github.com/digitalpalidictionary/dpd-db/assets/39419221/442c5deb-9946-442a-a487-47c44d4ddcfc)

**Reply by falko-strenzke** [reply]
That is one of the next goals, indeed. I thought about how to do it in the current state where the entries in the data base are just the plain HTML page. I could attempt to create the links during the parsing of the MDict file. It would probably make that process much more time consuming, though, since  now for every word in all the entries one database lookup is needed. I might give it a try.

The other option would be to wait until we have structured data in the database of the online  dictionary, that would make it easier because then then the link could be generated while rending the page.

**Comment by ghost** [comment]
Hello everyone. 

It is a great initiative. Thank you so much for the start. 

Is the dpd web interface code is available somewhere in github? 

http://niyamata.de/dict/dpd/search/

**Reply by falko-strenzke** [reply]
Yes, it is here: https://github.com/falko-strenzke/dpd-server-django

**Reply by ghost** [reply]
Thank you😊👍

**Comment by gambhiro** [comment]
Dear @falko-strenzke 

I made progress with integrating the DPD in [Simsapa](https://github.com/simsapa/simsapa) for dictionary search and rendering the results to HTML from the dpd sqlite file.

I was thinking that since you are using Python Django, perhaps calling some functions from Simsapa could be a way for you to run lookups and render the html.

The unit tests show what a lookup and rendering API looks like. For example [test_combined_search()](https://github.com/simsapa/simsapa/blob/main/tests/test_search_combined.py#L37) and [test_html_for_pali_words()](https://github.com/simsapa/simsapa/blob/main/tests/test_html_render_dictionary.py#L28).

I have been writing Simsapa with the assumption of a desktop app, but if this approach seems interesting for you, we could figure out a way to use it as a Python API from Django. In a server context, the database would have to be copied to the right place by hand but [it's not too difficult](https://simsapa.github.io/development/manual-database-setup/).

The large objects like AppData, TantivySearchIndexes, GuiSearchQueries etc. take long to initialize, but if they are attached to some object when Django starts, and the server routes can access them, then the responses should be decently fast.

These short screencasts show what the rendered HTML looks like and how the search is organized in the app.

[simsapa-ddpn-lookup-2024-01-07-20-36-49.mkv](https://drive.google.com/file/d/1Vsecacbk3i0ZAc_IObssXUt5sD0NQXOH/view?usp=drive_link)

[simsapa-lookup-and-links-2024-01-04-20-12-12.mkv](https://drive.google.com/file/d/1cGBlxSRmi8aO0WifUKd4dNVuUBWvGcXN/view?usp=drive_link)

**Reply by falko-strenzke** [reply]
Thank you for these hints, bhante. So far I had considered a potential approach for the linkification using Java Script. But I am not sure if and how well it would work. I will first checkout these two possible variants, that will still take me some time I think. If using the Simsapa functions is a good solution, I will come back to you :-)

**Comment by gambhiro** [comment]
@falko-strenzke In Simsapa, first I tried to identify specific parts of the text such as the grammar info, and add links with regex. This was fragile, often didn't catch the right words or created wrong links.

What has been more reliable is to hook into the `dblclick` event of the HTML page, get the selection and use that as a dictionary lookup query.

This mechanism also works when the user wants to double-click and lookup any other parts of the page, such as titles or non-Pali words (from translations for example).

**Comment by Devamitta** [comment]
done at https://dpdict.net/
